Assets/Scripts/Effects/EffectSpawner.cs

```diff
@@ -20,8 +20,19 @@ public class EffectSpawner : MonoBehaviour
     private void Start()
     {
         stateManager = this.EnsureComponent<PlayerStateManager>();
-        stateManager.CallOnStateEnter(triggerState, StateStart);
-        stateManager.CallOnStateExit(triggerState, StateEnd);
+        stateManager.OnStateChange += HandleNewPlayerState;
+    }
+
+    private void HandleNewPlayerState(State oldState, State newState)
```

Comment:
I think long-term, we should make `HandleNewPlayerState` an abstract method & make `EffectSpawner` a derived class (same for other classes w/this method).

Reply:
I generally try to stay away from inheritance for situations like these because c# doesn't let you do multiple inheritance. So let's say our base class here was "StateChangeListener". Now let's say we also wanted to have this component be a PunBehaviour or whatever it is called, then we are a little bit stuck